### DISCUSSION TITLE
Graph Query is very slow when fetching a tree

### DISCUSSION THREAD

**Discussion by BickelLukas** [ravendb/ravendb#13897] [Q&A]
I'm having problems querying a hierarchial tree using the graph query feature.

Our model is basically this:

```cs
class Activity {
  public string Id { get; set; }
  public string? ParentActivity { get; set; }
  ...
}
```

Given any Activity in the tree, I want to fetch the complete hierarchy to display it like this
``` 
- Activity 1
   - Activity 2
     - Activity 5
     - Activity 6
   - Activity 3
     - Activity 4
```

The RQL Query I came up with is this:
```
with {from Activities where id() == "activities/2-A"} as a

MATCH
// traverse back to the root
(a)
-recursive as _ (0, longest) { [ParentActivity]->(Activities) }
-[ParentActivity]->(Activities as parent)
or 
// recursively get all the children of root
(Activities as child)
-recursive as _ (0, all) { [ParentActivity]->(Activities) }
-[ParentActivity]->(parent)
```

I have added 500k activities to the database for testing (our customers will have something in the millions)  and the query is taking upwards of 10 seconds when executed. 

I have tried adding an index on `ParentActivity` but that didn't help. 
Is there any way to optimize this?

**Comment by ayende** [comment]
We are planning to remove graph queries from RavenDB, it didn't get the right pickup. See: https://ayende.com/blog/196673-A/request-for-comments-removing-graph-queries-from-ravendb?key=5e58086780354dd084bd5386fd5ee9e0
That said, you can usually achieve what you want using a standard RQL query. However, looking at the query, I'm not sure that I'm following what you meant to do.

This will go from `2 -> 1` up, (roots).
But then it will *scan all the documents you want* and give you all of them that end up with the root.

The result of this query is _all documents in the database_, no?
That can be seriously expensive and not surprising that it takes a toll. 

Is that what you meant to do? Can you give me some insight into what it is that you are trying to do?

**Reply by BickelLukas** [reply]
Thats to bad, we are currently evaluating Database technologies and the contenders are: ArangoDb, MarkLogic and RavenDb

I would like to go with RavenDb because of the great .Net client and the graph feature is the only thing missing that the other ones have. 

But we don't necessarily need graph if we can get the desired results with great performance some other way.

So to clarify what we're trying to do:
Yes, we do need to return all the documents that end up at the root. But no, this would not return the complete database.

The ParentActivity property is optional so we have mutliple roots and mutliple trees of activities. 

You can imagine an Activity as being an E-Mail and the Tree representing a whole conversation. (a Answer would have the original message as its parent)

These trees usually don't get any bigger than about 100 Activities.

**Reply by vlko** [reply]
As for Tree structures, you can easily implement loading tree with _select output()_ javascript function with iterating over parent/children nodes and loading by id with _load(document_id)_. We implemented such as scenario and it takes tens of ms to process tree with hundreds up to tousand nodes.
Cons is that such as output function is pure javascript, not some nice C# code.

**Reply by BickelLukas** [reply]
In case anyone else is wondering the same thing. With the input of @vlko I was able to create the following.

First a Map/Reduce Index : 
```cs
public class Activity_Children : AbstractIndexCreationTask<Activity, Activity_Children.Result>
{
    public class Result
    {
        public string Parent { get; set; }
        public List<string> Children { get; set; }
    }

    public Activity_Children()
    {
        Map = activities =>
            from act in activities
            where act.ParentId != null
            select new Result
            {
                Parent = act.ParentId,
                Children = new List<string>() { act.Id }
            };

        Reduce = result =>
            from act in result
            group act by act.Parent into g
            select new Result()
            {
                Parent = g.Key,
                Children = g.SelectMany(x => x.Children).ToList()
            };

        OutputReduceToCollection = "Activities/Children";
        PatternForOutputReduceToCollectionReferences = result => result.Parent + "/Children";
    }
}
```
This creates a Activities/Children collection that is basically the inverse mapping of the reference property on the activity.

Then I can query the history tree using the following javascript query:
```js
declare function getHistory(act) {
    let root = act;
    while (root.ParentId) {
        root = load(root.ParentId);
    }
    
    return getChildren(root);
}

declare function getChildren(act) {
    const hierarchies = [];
    const activities = [ act ];
    
    const reference = load(id(act) + '/Children');
    if (reference) {
        const childrenDocument = load(reference.ReduceOutputs[0]);
        
        hierarchies.push({ id: id(act), children: childrenDocument.Children })
        
        for (let i = 0; i < childrenDocument.Children.length; i++) {
            const child = load(childrenDocument.Children[i]);
            
            const nested = getChildren(child);
            hierarchies.push(...nested.hierarchies);
            activities.push(...nested.activities);
        }
    }
    
    return { hierarchies, activities };
}

from "Activities" as a where id(a) == "activities/2-A" select getHistory(a)
```


So far the query seems to perform well.

**Comment by ayende** [comment]
Given the scenario you have, the _manner_ in which you model is is not a good one.
You are hammering at the solution, but there is no real need for that. Instead, you can do the following:

* `Activities` collection (Has `ParentActivity`)

What you want to get as a result of the query is all the activities belong to the same `Bundle` (for lack of a better context).
That means that you can add a `BundleId` to your documents and then query on that. The end result is the same (all the relevant documents), but the cost of the query and the modeling constraints are far simpler.
